### ISSUE TITLE
defining an open, collaborative process of merging our efforts together

### ISSUE THREAD

**Issue by eighthave** [ibauersachs/otr4j#15]
After an introductory email thread (that unfortunately is private), I think its time to focus in on the specific issue of defining how we all work together on otr4j.  Most importantly, this is to make sure that all of our work stays in a single, shared repository used by all of the projects.

First and foremost, we need a light and effective structure, we do not need a Constitution like Debian ;-). This is a project that we all contribute to in little bits and spurts.  Here's what I propose, based on the email discussion:
- any developer can request push access, regardless of project or organization affiliation
- _all_ contributors submit code via pull requests
- new commits must be pushed by the reviewer of the pull request, not the author
- "lazy consensus" approach for granting push access: 
  - anyone with push access can vote/veto
  - if about a week or so as passed after requesting push access and no one has objected, then that requester can be granted push access

I'm a big fan of doing this all using git remotes.  I have documented how we set up this process here:
- https://guardianproject.info/2013/11/21/a-tag-team-git-workflow-that-incorporates-auditing/

Here are the projects/orgs/people that I know of that are using otr4j:
- ChatSecure - @guardianproject @devrandom @n8fr8 @eighthave
- Conversations - @siacs
- Jitsi - @jitsi @gpolitis @mdzhigarov 
- Xabber - @redsolution @a-iv
- other contributors that I don't know which projects they are affiliated with: @tribut @cobratbq @fmanco

Please ping anyone else you think should be on this thread.

**Comment by gpolitis** [comment]
+1 I personally love this approach.

**Comment by tribut** [comment]
Not that I feel my voice should carry much weight, but this seems like a sensible approach. :+1:

**Comment by n8fr8** [comment]
Sounds good to me. Hope we can get everyone on board for this, under the spirit of mutually beneficial compromise!

**Comment by devrandom** [comment]
+1

**Comment by fmanco** [comment]
I did some bug fixes/reports back when I was a student at University of
Aveiro. I'm not currently working with otr4j, and I probably won't do so in
a near future. Anyway it sounds a good plan, keep up the good work.

# 

_Filipe  Manco_
me.fmanco.org

2014-12-15 18:55 GMT+00:00 Felix Eckhofer notifications@github.com:

**Comment by cobratbq** [comment]
```
+1

  And for completeness: I'm @cobratbq. I'm a committer with Jitsi
  and I implemented fragmentation support for outgoing messages.

  I'm also interested in keeping otr4j generic enough so that it can
  be used for other protocols, such as IRC. Just wanted to point
  that fact out in a (largely) XMPP crowd ;-)

  On a side-note: I have a personal branch where I prepared a
  simplification of outgoing fragmentation support. Basically, I
  overlooked something and made the implementation slightly too
  complex. I hadn't merged it yet, because I wanted to do a review
  before merging. I try to be careful with these kinds of "quick"
  and "simple" changes. I've fixed that in
  https://github.com/cobratbq/otr4j/tree/otr-simplify so that'll be
  up and coming in the next few days.



  On 15-12-14 15:24, Hans-Christoph Steiner wrote:



  After an introductory email thread (that unfortunately is
    private), I think its time to focus in on the specific issue of
    defining how we all work together on otr4j. Most importantly,
    this is to make sure that all of our work stays in a single,
    shared repository used by all of the projects.
  First and foremost, we need a light and effective structure, we
    do not need a Constitution like Debian ;-). This is a project
    that we all contribute to in little bits and spurts. Here's what
    I propose, based on the email discussion:

    any developer can request push access, regardless of project
      or organization affiliation

      all contributors submit code via pull requests
    new commits must be pushed by the reviewer of the pull
      request, not the author
    "lazy consensus" approach for granting push access:

        anyone with push access can vote/veto
        if about a week or so as passed after requesting push
          access and no one has objected, then that requester can be
          granted push access



  I'm a big fan of doing this all using git remotes. I have
    documented how we set up this process here:

    https://guardianproject.info/2013/11/21/a-tag-team-git-workflow-that-incorporates-auditing/

  Here are the projects/orgs/people that I know of that are using
    otr4j:

    ChatSecure - @guardianproject @devrandom @n8fr8 @eighthave

    Conversations - @siacs

    Jitsi - @jitsi
      @gpolitis @mdzhigarov 
    Xabber - @redsolution
      @a-iv

    other contributors that I don't know which projects they are
      affiliated with: @tribut
      @cobratbq @fmanco


  Please ping anyone else you think should be on this thread.
  —
    Reply to this email directly or view it on
      GitHub.
```

**Comment by eighthave** [comment]
So the next step is getting people added as committers.  This also is required before issues and pull requests can be assigned to people. I'm hereby requesting commit access.  How shall we handle this process?  Who can add committers to this repo?

**Comment by gpolitis** [comment]
I don't have the admin rights to do so. @emcho I know that you're pretty busy this period, but as an owner of this project, could you please chime in?

**Comment by emcho** [comment]
Hey folks,

Good to see all the enthusiasm! The process of obtaining commit access to Jitsi projects is described here: https://jitsi.org/commitaccess

**Comment by gpolitis** [comment]
I think this mostly affects the "lazy consensus" approach for granting push access. I should have known better and point you to that page earlier.

**Comment by ibauersachs** [comment]
@emcho do we consider otr4j really a Jitsi project?

**Comment by emcho** [comment]
I am not sure what you mean. In what respect is it not one?

**Comment by gpolitis** [comment]
@ibauersachs this should probably be part of a `ChangeLog` file or something, but, in absence of such a file, maybe a brief history reminder would be helpful.

I originally developed otr4j for Jitsi (SIP Communicator) during GSoC 09 under jitsi.org [mentorship](https://jitsi.org/GSOC2009/OTR). The code was hosted [here](https://code.google.com/p/otr4j) for some time. At some point, I wanted to move the project over to GH to take make it easier for people to share their improvements and/or their bug fixes (with pull requests and stuff…).

Jitsi wasn't present at GitHub and @a-iv had already moved the project there and had changed some things, which was great and I acknowledge his contribution in the THANKS.markdown file. So, instead of recreating the project, I simply forked @a-iv’s one to gpolitis/otr4j.

Things have changed, jitsi moved to GH and, so some time ago, I moved the project (again ...) under the jitsi organisation.

I actually regret having moved the project so much, as it keeps causing confusion to this day. There was nothing wrong with Google Code to begin with, so if I were to take a decision now, I’d leave it on Google Code.

**Comment by ibauersachs** [comment]
Well I knew that you started it as a GSoC project under Jitsi's umbrella, but not all the intermediary steps. But they only make me feel more the way that it is an independent project - currently hosted @Jitsi. (Some of) The contributions you mention probably don't even have the BlueJimp's CLA signed.

**Comment by gpolitis** [comment]
As I said, I regret having moved the project around so much, as it keeps causing confusion to this day. I never had any intension to move the project away from Jitsi though.

The bulk of the code and most contributions are from within the Jitsi community (EDIT: I'm speaking for this version of otr4j). It's true that there are two contributors that haven't signed the BlueJimp's CLA, but I didn't know that such a thing even existed a couple of months ago.

**Comment by emcho** [comment]
@ibauersachs regardless of whether or not I agree with your feeling about the situation, I don't think we are having the right debate. The question I was answering from Hans was: "What do I need to in order to get commit access to this project". The answer to that stays. 

I know that the ChatSecure guys have already forked that code and using that fork (which is obviously totally fine) so no one is chained to using the Jitsi instance of OTR4J.

In the same way there is no reason for Jitsi to move to another fork either as we are perfectly fine with OTR4J (the instance under Jitsi) and I see why that project would be treated differently to anything else we have, or why we would want to move it out of there.

**Comment by ibauersachs** [comment]
The very reason this issue exists is to unfork the project and combine the efforts. And Hans' question refers to what the issue creator proposed and the resulting discussion/votes.

**Comment by eighthave** [comment]
@emcho if your requirement for a Contributor License Agreement (CLA) is inflexible, then I'm afraid we'll have to take our work a separate repo (i.e. otr4j/otr4j), one that does not have a CLA where anyone is free to join without signing something.  I think that a CLA goes contrary to an open, collaborative project since it reserves special rights for one set of contributors, and the process puts legal drudgery onto the contributors.  You will also need to review the otr4j code contributions since it is likely that not all contributions came from people who signed your CLA (e.g. code from java-otr).

**Comment by eighthave** [comment]
If you really need a CLA, Apache's seems very reasonable, and has the benefits of being an open standard like the LGPL license used in otr4j (rather than a custom, BlueJimp legal document) 
https://www.apache.org/licenses/icla.txt

**Comment by gpolitis** [comment]
I'm afraid that debating the CLA here will result in a stalemate. Also, I believe the right place to have this discussion is on the jitsi-dev mailing list and not here.

I'm still in favor of all @eighthave's proposals, except for the last one that is not compatible with the procedure for becoming a Jitsi committer.

**Comment by n8fr8** [comment]
I agree with @gpolitis, in that we don't have the bandwidth currently to resolve the diplomatic and corporate issues on this front. What I propose we do is to create our own "community" branch at otr4j/otr4j which we will keep up to date with jitsi/otr4j and that will not require a CLA to contribute to. Any enhancements or fixes we implement will be available to jitsi under the terms of the APL. Just as I am sure Jitsi has done in other cases, you can choose at that point, whether or not you want to include third-party open-source code into your codebase. For other app developers, they can decide on their own whether they want our enhancements or not, and ideally, not have to make any other trade-offs regarding stability or up-to-dateness.

Perhaps in six months, we can revisit the issue of the CLA, and review the progress we have made?

**Comment by languitar** [comment]
+1 for the general idea.

I am also in favor of having a repository, where contributing is not hindered by requiring a CLA.

**Comment by gpolitis** [comment]
@languitar I believe the right place to debate the BlueJimp CLA is on the jitsi-dev mailing list and not here. Apologies for the inconvenience of having to sign the CLA but it's a requirement across all BlueJimp projects.

To move things forward, I have updated the README.markdown file with some guidelines on how to contribute to otr4j. The guidelines are based on @eighthave suggestions for the most part, with the following two modifications: 
- I added to the 2nd bullet the requirement about the [BlueJimp CLA](http://bluejimp.com/bca.pdf) (again, apologies for the inconvenience of having to sign it).
- I replaced the last bullet with the [current approach](https://jitsi.org/commitaccess) for granting push access.

**Comment by gpolitis** [comment]
@n8fr8 thank you for letting us know of your plans regarding otr4j/otr4j. Everything is possible, so maybe something changes with respect to the CLA. With that in mind, it sounds like a good idea to revisit the situation in six months.

I'm sorry we couldn't fully resolve the issue now but thank you for bringing it up :)

**Comment by ibauersachs** [comment]
@gpolitis Maybe put the contributing section into a [CONTRIBUTING.md](https://github.com/blog/1184-contributing-guidelines) as well. We probably should do something like that for all the projects...

**Comment by emcho** [comment]
Hey guys,

Sorry for the lag. Busy times. Our CA is the same as the one used by the jxta and (at one point) openoffice projects. It is disttributed under the CC license (not that I see how this is relevant).

In the past we have also accepted that contributors simply release their patches in the public domain ... although this is much more hassle since it needs to be clearly stated for every commit.

If either of those is a problem for you then, indeed, continuing your fork sounds like the only alternative.

Emil

**Comment by eighthave** [comment]
Jumping back into this thread with the hopes of resolving this.  The code from @devrandom and from the original java-otr was definitely not covered by the CLA, and potentially other contributors' code as well.  That also means that the license conversion from LGPL to Apache is not valid since those authors are still the copyright holders.  

How about we just work with the existing reality that otr4j is not covered by the bluejimp CLA as a whole, and make a single, collaboratively maintained otr4j?

And FYI, I'm not opposed to converting the license from LGPL to Apache-2.0, but that does require the approval of all copyright holders.

**Comment by tribut** [comment]
My code was contributed under LGPL and the headers still state that it was: https://github.com/jitsi/otr4j/blob/master/src/main/java/net/java/otr4j/session/OtrAssembler.java

I don't think I have signed a CLA for jitsi. I have no general reservations regarding the Apache license but the way this is handled here doesn't really encourage me to sign anything "ex post"...

**Comment by gpolitis** [comment]
If there are problems with some parts of the code with respect to licensing, I think I'd rather rewrite those parts. @eighthave @devrandom, we have switched from the bluejimp CLA to the Apache-2.0 CLA and I remember you guys were open to that. Do you think there's a chance for you to reconsider the fork and merge the efforts here?

**Comment by tribut** [comment]
So just to be clear on the process: The thing [mentioned in the README](http://bluejimp.com/bca.pdf) is no longer relevant? The contribution guidelines for the main jitsi repo mention https://jitsi.org/icla - that's the current one?

**Comment by gpolitis** [comment]
Hi @tribut, you're right, the README file is outdated.

**Comment by cobratbq** [comment]
Okay, it's obvious that you're welcoming CLA'ed contributions. What bothers me is that you don't seem very concerned with otr4j. To me, it feels like for you it's just another project that needs to be maintained. The latest release was published on 2015-02-22.

For example, when I found the SMP boundary checking bugs, I let you know. I then replied to your response to let you know that I cannot do the otr4j Sonatype repo publishing + asking for whether we should update people on the mailinglist + asking if you had any idea on whether exploitation was possible. I never got a response. (For reference, I'm referring to commit https://github.com/jitsi/otr4j/commit/7c67c7a62f80a90e71191825f8abbdf5412b6dd1 which fixed some protocol boundary checking which wasn't according to spec.) Now, I'm not saying that this was a big deal per se. I'm no crypto expert and I haven't been able to figure out whether or not it is, myself. But what does not give me much confidence is the passivity surrounding this in general.
